### ISSUE TITLE
R&D on state preservation for simpleLayers, user added services, and files

### ISSUE THREAD

**Issue by milesap** [fgpv-vpgf/fgpv-vpgf#2702]
Examine and propose some methods of preserving the state of simpleLayers, user added services, and files. The end goal is to have each of these be transformed into a string or JSON for cloud storage and the new bookmark functionality.

Some initial thoughts:
- Can user added services be modified so that it generates a config snippet? 
- Can we convert all user imported files and related customizations via the UI interface into a standard format like geoJson?
- `simpleLayer` doesn't seem simple... not sure serializing the class instance would work as it's tightly coupled with geoApi. Might need to add `toJson` and `fromJson` to the class and all geometry.

**Comment by james-rae** [comment]
> Can user added services be modified so that it generates a config snippet?

Yes.  I believe there is already a "snippet" being generated in the wizard; it likely gets passed to the loader and then tossed. Just need to grab it / store it / manage it.

> Can we convert all user imported files and related customizations via the UI interface into a standard format like geoJson?

This is already happening in the geoApi.  Can also consider storing the `EsriJson` format as well; would save us re-processing the `geoJson` format, though we would need a new endpoint to process `EsriJson`. We would also need to consider capturing the colour the user chooses for symbolizing the layer.

We may also need to impose size limits for our storage.  We could add a warning icon to the legend or the wizard that informs a user that if a file is over a certain size, it will be lost on projection/language/bookmark/etc.

> `simpleLayer` doesn't seem simple...

Agreed, probably something like serializing the geometry is most straightforward.

**Comment by milesap** [comment]
I wonder if there is a need to preserve a `simpleLayer`? Any plugins that make a `simpleLayer` or draw on the maps `simpleLayer` can store custom info in the new state store and just recreate the changes when the map is reloaded.

**Comment by milesap** [comment]
The plugin that allows the user to draw (the circles) would need to store the geometry lat/lngs or w/e in the state store and re-draw them on map reload. A little more work for them however - we can always add state preservation for simpleLayers at a later time.

**Comment by james-rae** [comment]
> we can always add state preservation for simpleLayers at a later time

Sounds like a prudent choice to postpone and re-examine later  :white_check_mark:
